### PR TITLE
[inductor] benchmark fusion

### PR DESCRIPTION
**Pull request by shunting314**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #108193

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @Xia-Weiwen @ngimel @anijain2305

### CODE REVIEW

**Comment by shunting314** [comment]
benchmark fusion helps with huggingface [link](https://hud.pytorch.org/benchmark/huggingface/inductor_with_cudagraphs?startTime=Wed%2C%2023%20Aug%202023%2019%3A09%3A47%20GMT&stopTime=Wed%2C%2030%20Aug%202023%2019%3A09%3A47%20GMT&granularity=hour&mode=training&dtype=amp&lBranch=shunting-benchmark-fusion&lCommit=ecc85db2c1559abbf12f6dcb2bc05c1734faff02&rBranch=main&rCommit=138e2895d08a6517c5718b2a0118c1b23ff4664c) . A few things worth mention
1. a few models fails, I'll need debug
2. benchmark fusion will increase computation time. So I'll disable it by default.

**Comment by shunting314** [comment]
New round of perf run [link](https://hud.pytorch.org/benchmark/compilers?startTime=Fri%2C%2025%20Aug%202023%2017%3A07%3A24%20GMT&stopTime=Fri%2C%2001%20Sep%202023%2017%3A07%3A24%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=shunting-benchmark-fusion-4&lCommit=51b831f9c3adf28a651575c7b9f82e30bea7a373&rBranch=main&rCommit=3a79621c9dce17f77fbddc06aab21f6bc477f313)
- Perf 
  - TB 1.69x -> 1.73x
  - HF: 1.78x -> 1.81x
  - TIMM: 1.69x -> 1.70x
- Compilation time increase as expected: increase by about 3x. So this feature will be disabled by default.
- Compression ratio drops by about 10%. I think it's a side effect of changed fusion: A pair of nodes whose fusion attempt is rejected may later on be fused with nodes in a longer distance.
- There is a single remaining accuracy failure for HF I need debug, but that should not affect the overall metric much.

**Comment by jansel** [comment]
Awesome to see these perf improvements.

1) We should enable this in max-autotune mode

2) Are there any heuristics we could take from this to improve things without autotuning?

**Comment by lezcano** [comment]
A side note: Compiled triton kernels already expose whether there were spilled registers and allocated registers under `CompiledKernel.{n_spills,n_regs}`. These could potentially be used as hints for whether to continue fusing stuff or whether to stop and even recompile a kernel in the next execution.

File: torch/_inductor/codegen/triton.py [hunk 1]
```diff
@@ -2729,6 +2747,68 @@ def select_tiling(cls, node_schedule, numel, reduction_numel=sympy.Integer(1)):
...
+        for n in nodes:
+            n.last_usage = set()
+
+        self.codegen_node_schedule_with_kernel(node_schedule, kernel)
+        with config.patch("benchmark_kernel", True), V.set_kernel_handler(kernel):
+            src_code = kernel.codegen_kernel()
+        mod = PyCodeCache.load(src_code)
+
+        def cache_file_path():
+            return os.path.splitext(mod.__file__)[0] + ".kernel_perf"
```
Comment (aakhundov) [inline torch/_inductor/codegen/triton.py:2775]: Do you think it would be possible to rely on the `AlgorithmSelectorCache` here instead (not necessarily in this PR, but maybe in a follow-up), the same as the autotuning currently relies on? Then we could automatically reap the benefits of autotuning improvements: centralized caching, parallelized autotuning (currently WIP), managed caching (e.g., for internal / prod use cases). Also, I believe, that would open the door for easier "autotuning + epilogue fusion" integration.

**Comment by aakhundov** [comment]
> Compilation time increase as expected: increase by about 3x. So this feature will be disabled by default.

@shunting314 Perhaps, instead of a binary flag of either benchmarking *all* possible fusions or *none*, we could introduce some kind of multi-level configuration flag? E.g., in our experience with large models, epilogue fusion can lead to serious perf regression. So maybe we could have an intermediate config / flag value to benchmark only epilogue fusions? That may result in (much) faster than 3x compilation time, but still guard from important perf regression.

**Comment by aakhundov** [comment]
> E.g., in our experience with large models, epilogue fusion can lead to serious perf regression.

More context on this. The script below is a minimal repro extract from a larger model (the shapes are real, the graph is similar). What it does is MM + epilogue, fused and unfused, and compares the runtimes:

```python
import torch

from torch._inductor import config
from torch._inductor.utils import do_bench


def mm(a, b, c, d, e, f, g):
    x = a @ b
    y1 = torch.sigmoid(x + c) * g
    y2 = torch.sigmoid(x + d) * g
    z1 = torch.sigmoid(x + e) * g
    z2 = torch.sigmoid(x + f) * g
    return y1 + y2, z1 + z2


inputs = [
    torch.rand(1024, 5098, dtype=torch.half, device="cuda"),  # a
    torch.rand(5098, 1024, dtype=torch.half, device="cuda"),  # b
    torch.rand(1024, 1024, dtype=torch.half, device="cuda"),  # c
    torch.rand(1024, 1024, dtype=torch.half, device="cuda"),  # d
    torch.rand(1024, 1024, dtype=torch.half, device="cuda"),  # e
    torch.rand(1024, 1024, dtype=torch.half, device="cuda"),  # f
    torch.rand(1024, 1024, dtype=torch.half, device="cuda"),  # g
]

runtimes = []
for epilogue_fusion in [True, False]:
    with config.patch({
        "max_autotune": True,
        "max_autotune_gemm_backends": "TRITON",
        "epilogue_fusion": epilogue_fusion,
    }):
        torch._dynamo.reset()
        mm_with_fusion = torch.compile()(mm)
        out_with_fusion = mm_with_fusion(*inputs)
        runtime = do_bench(lambda: mm_with_fusion(*inputs))
        runtimes.append(runtime)

t_with_fusion, t_no_fusion = runtimes

print(f"runtime with fusion: {t_with_fusion:.3f} ms")
print(f"runtime w/o fusion: {t_no_fusion:.3f} ms")
print(f"slowdown from fusion: {t_with_fusion / t_no_fusion:.0f}x")
```

Running the script on A100 80GB produces this output:

```
runtime with fusion: 3.059 ms
runtime w/o fusion: 0.113 ms
slowdown from fusion: 27x
```

My understanding is that, due to excessive use of RF in the fused mm + epilogue version, the resulting Triton kernel is running into register spilling. This is to motivate introducing an intermediate configuration / flag to benchmark epilogue fusions only.

**Comment by shunting314** [comment]
> Awesome to see these perf improvements.
> 
> 1. We should enable this in max-autotune mode
> 2. Are there any heuristics we could take from this to improve things without autotuning?

For 1, I tried this on GPT2ForSequenceClassification to observe the impact on compilation time (note: benchmark fusion does not speedup this model much as the dashboard shows. We mainly focus on compilation time here). Here are the results:

(CDT for coordinate descent tuning, BMFusion for benchmark fusion. The last column represents compilation time in seconds. All experiments are run with an empty cache.).

Baseline (OOB)
- 2.291006,40.933383,82.973447 (1x)
MaxAutotune
- 2.353730,39.960711,198.890689  (2.40x)
MaxAutotune + CDT
- 2.369012,39.948145,229.139282 (2.76x)
MaxAutotune + BMFUSION
- 2.358858,39.678820,260.968196 (3.15x)
MaxAutotune + CDT + BMFUSION
- 2.363157,39.560971,425.247536 (5.13x)

Some observations below:
- Enable BMFusion upon MaxAutotune increase compilation time from 2.4x -> 2.76x. Looks not too bad compared to MaxAutotune.
- Enable MaxAutotune + CDT + BMFUSION will cause 5.13x slow down of compilation. The main reason is some kind of compound effect: if we enable all those options, when we benchmark a kernel to make fusion decision, we need do MaxAutotune and CDT to do relatively heavy search for promising triton configs.

Go back to the review comment, enable BMFusion by default when MaxAutotune is on looks not too bad based on this experiment. But we still increase compilation time. Should it be a concern that users of MaxAutotune will suddenly see increased compilation time with the same command? (although they may also see speedup).

**Comment by shunting314** [comment]
For `Are there any heuristics we could take from this to improve things without autotuning?`

yea, I'll try to find some heuristics. register spilling mentioned by @lezcano  is one way and we already leverage this in coordinate descent tuning. I'll also try to gather more examples so we have more chance to find more patterns.

File: torch/_inductor/codegen/triton.py (hunk 1 above)
Comment (shunting314) [reply torch/_inductor/codegen/triton.py:2775]: Hmm, `AlgorithmSelectorCache` has a structure designed for MM tuning: e.g. tensor shapes and benchmark results are stored in the cache. While benchmark fusion here should be generally applicable to pw/reduction/template. So I don't find it straightforward to directly reuse `AlgorithmSelectorCache` here, unless you see some good ways 

On the other hand, both AlgorithmSelectorCache and the code here rely on `CodeCache` class. So I think a `managed caching` (I assume this as a cache shared by multiple machines) can still be done in `CodeCache` level.

Also the workload here is not as easy parallelizable as mm tuning. For mm tuning, we know before handle a bunch of triton configs to parallelize. Here we don't. But I can image with some redesign, we can batch and parallelize the benchmarking requests.

**Comment by shunting314** [comment]
> @shunting314 Perhaps, instead of a binary flag of either benchmarking all possible fusions or none, we could introduce some kind of multi-level configuration flag? E.g., in our experience with large models, epilogue fusion can lead to serious perf regression. So maybe we could have an intermediate config / flag value to benchmark only epilogue fusions? That may result in (much) faster than 3x compilation time, but still guard from important perf regression.

Yea, agree. The PR does not support this ATM. For pointwise/reduction, our codegen already support generating harness for benchmarking (e.g. create random inputs), for TemplateKernel, we don't have such harness generated yet. But this should not be hard. I'll do that in a followup:  https://github.com/pytorch/pytorch/issues/108716

**Comment by shunting314** [comment]
>@shunting314 Perhaps, instead of a binary flag of either benchmarking all possible fusions or none, we could introduce some kind of multi-level configuration flag? E.g., in our experience with large models, epilogue fusion can lead to serious perf regression. So maybe we could have an intermediate config / flag value to benchmark only epilogue fusions? That may result in (much) faster than 3x compilation time, but still guard from important perf regression.

> Yea, agree. The PR does not support this ATM. For pointwise/reduction, our codegen already support generating harness for benchmarking (e.g. create random inputs), for TemplateKernel, we don't have such harness generated yet. But this should not be hard. I'll do that in a followup: https://github.com/pytorch/pytorch/issues/108716

BTW, even if we have done this, the mm epilogue fusion problem you mentioned is only partially solved. Since at the fusion stage, we already baked in the block size in the triton template kernel. The best we can do here is to not fuse too much stuff in the epilogue which may slow down the kernel. But fundamentally, we should delay the decision of block size until we try to do fusion. This way we may pick a different block size which can be more suitable for fusion.

**Comment by lezcano** [comment]
The block fusion may help, if you have fused too much stuff already to a large mm (or any compute-heavy op, really) you're going to deteriorate the performance.

In https://github.com/pytorch/pytorch/pull/90994, I wrote an algorithm that would give an upper bound on the number of registers an operation would use to try to estimate a better block size. The benchmarks said this introduced regressions, and I ended up dropping it, but it may be relevant here. If we just do epilogue fusion of pointwise ops, estimating the number of registers shouldn't be too difficult for those ops.

In general, I guess that you wouldn't want to fuse anything that increases the number of regs for an `mm`, or at most allow for a teeny tiny bit of increased register pressure.

File: torch/_inductor/codegen/triton.py (hunk 1 above)
Comment (aakhundov) [reply torch/_inductor/codegen/triton.py:2775]: Thanks for the clarification, makes sense. The reason I've asked about unification with the `AlgorithmSelectorCache` is that I was imagining that, at a later point, we could defer autotuning to be done in an epilogue fusion-aware way: e.g., select the best option from autotune choices with and without epilogue fusion. And then there would be plently of parallelism, too. This would likely require bringing `AlgorithmSelectorCache` into the picture started in this PR, but maybe it's too abstract at this point.

**Comment by aakhundov** [comment]
> But fundamentally, we should delay the decision of block size until we try to do fusion. This way we may pick a different block size which can be more suitable for fusion.

**Comment by shunting314** [comment]
> In https://github.com/pytorch/pytorch/pull/90994, I wrote an algorithm that would give an upper bound on the number of registers an operation would use to try to estimate a better block size. The benchmarks said this introduced regressions, and I ended up dropping it, but it may be relevant here. If we just do epilogue fusion of pointwise ops, estimating the number of registers shouldn't be too difficult for those ops.

@lezcano  interesting to see we already did some work to estimate the register usage. Although it would be more relevant to have the estimation of spilled registers which also depends on block size and num-warps? Also if we can afford compiling the kernel, we should already get the accurate register usage from ptxas

**Comment by shunting314** [comment]
> If we keep the entire set of choices unresolved till the benchmark fusion is done, when exactly are we going to make the choice from the set?
@aakhundov I think
1. if people could not afford the increased compilation time, they can enable just one of 'deferred autotune' or 'benchmark fusion' rather than both
2. if they enable both for better performance, here is how I imagine how it should work: whenever we have a fusion choices to fuse a matmul (possible to already have a non-empty epilogue) and some extra epilogue nodes, we need benchmark the fused kernel . That benchmark run will trigger max-autotune we do before: i.e., we try various choices of block sizes for the fused matmul.  This indeed increase compilation time. But I think we do epilogue fusion after other fusions by default, so hopefully we don't benchmark a matmul with an epilogue too often (since the epilogue is itself an already fused node).

File: torch/_inductor/codegen/triton.py (hunk 1 above)
Comment (aakhundov) [reply torch/_inductor/codegen/triton.py:2775]: Actually, `AlgorithmSelectorCache` relies of `PersistentCache`, while in this PR `PyCodeCache` is used. My understanding is that those are disjoint. So maybe we'll need to consider a managed `PyCodeCache` location, too.

**Comment by aakhundov** [comment]
> whenever we have a fusion choices to fuse a matmul (possible to already have a non-empty epilogue) and some extra epilogue nodes, we need benchmark the fused kernel . That benchmark run will trigger max-autotune we do before: i.e., we try various choices of block sizes for the fused matmul.

This sounds reasonable. However, I'm still missing one thing: at some point during this process (which, for a single template buffer, can happen multiple times due to multiple parallel `Pointwise` epilogue fusion opportunities) we need to select the winner of the autotuining and stop considering all possible choices. When do you think this selection should happen? One option may be at the first template + pointwise fusion decision. Another: when deferred autotuning is enabled, just returning `True` for all `can_fuse(template, pointwise)` calls, and then at the very end do a holistic benchmarking (not sure what this means, but we'd have all necessary information in place for a global decision). What do you think?

**Comment by jansel** [comment]
Feel free to re-request review once this is ready for me to take a closer look.

File: torch/_inductor/codegen/triton.py (hunk 1 above)
Comment (shunting314) [reply torch/_inductor/codegen/triton.py:2775]: Good catch yea. Maybe we need some refactoring to do the decision of cache location in some sort of base class

**Comment by shunting314** [comment]
> This sounds reasonable. However, I'm still missing one thing: at some point during this process (which, for a single template buffer, can happen multiple times due to multiple parallel Pointwise epilogue fusion opportunities) we need to select the winner of the autotuining and stop considering all possible choices. When do you think this selection should happen? One option may be at the first template + pointwise fusion decision. Another: when deferred autotuning is enabled, just returning True for all can_fuse(template, pointwise) calls, and then at the very end do a holistic benchmarking (not sure what this means, but we'd have all necessary information in place for a global decision). What do you think?

@aakhundov I was thinking every time we need decide if we should fuse a template buffer with an epilogue, we can run through all the choices for matmul autotuning. The reason is the winning choice may be different depending on the epilogue we fuse. If compilation time is too long for a model/use case, we can disable benchmark fusion, which basically is the second option you mentioned.

**Comment by shunting314** [comment]
I gathered the whole list of slow fusions for HF leverage PR (https://github.com/pytorch/pytorch/pull/109245) : https://gist.github.com/shunting314/964e73cc98368b301414ec7b7ad4c702 . 

I went thru some of the slowed down fusion with >1ms absolute latency for the fused kernel. One pattern pop up: Both kernel 1 and kernel 2 uses around 32 registers per thread and the fused kernel uses around 64 registers per thread. The fusion will impede occupancy and cause slow down.  
Example kernels:
- kernel 1: https://gist.github.com/shunting314/8317d702c1acb1b3501df6cd6a6935ad
- kernel 2: https://gist.github.com/shunting314/ef2138f7b40ce6e018774f0a94ef1f51
- fused kernel: https://gist.github.com/shunting314/0726d1d599aee478902d8d32d9f533ae
Slowed fusion caused by this pattern (register pressure) can be improved by reducing RBLOCK. I make the improvement in a standalone PR (https://github.com/pytorch/pytorch/pull/109275)

I'll look more into the slow fusions. We should get more insights.

**Comment by jansel** [comment]
Interesting, maybe we need to write something to estimate register usage (without an expensive compile) to use in our heuristics.

**Comment by shunting314** [comment]
> Interesting, maybe we need to write something to estimate register usage (without an expensive compile) to use in our heuristics.

yea, the most complex part here is probably doing register allocation if we want to estimate the register usage? Or maybe a simple algorithm is fine if we only need a (relatively loose) upper bound

**Comment by shunting314** [comment]
2 more patterns I found from the bad fusions collected for HF.

# Pattern 1: register spilling

Fuse https://gist.github.com/shunting314/cd4dea7cbd337ae3939e9e1da270e263
and https://gist.github.com/shunting314/14035e31399a406297ed4f531d22bdf6
we get kernel https://gist.github.com/shunting314/02a9be7a59da482a9fdfc45e33cb9d91 
which has register spills.

kernel1: n_regs 111, n_spills 0, latency 0.23ms
kernel2: n_regs 48, n_spills 0, latency 0.023ms
fused kernel: n_regs 255, n_spills 94, latency 1.93ms

I can update this PR like: if we found any kernel spill registers, we skip the fusion.

# Pattern 2: inaccurate memory bandwidth saving estimate for fusion

Consider 2 kernels.
```
Kernel1: 
for ...:
  load(in_ptr1)
```

kernel2:
```
for ...:
  load(in_ptr1)
```

Fusing kernel1 and kernel2 will like:
```
for ...:
  load(in_ptr1)

for ...:
  load(in_ptr1)
```

Note the 2 loops need to be separate if the second one need the first reduction's result. (Like softmax)

In our fusion code, we would assume we save the access to `in_ptr1` once, but actually there is no savings since in the fused kernel, we need access in_ptr1 twice.

I'll create separate PR to improve our code that estimate memory bandwidth saving for fusion.

**Comment by lezcano** [comment]
It's odd that those kernels, when fused, have so many spilled registers, when the fusion is pretty much concatenating the code from one after the other... I wonder whether triton keeps some registers alive for longer than it should just because we are marking the loads with`evict_last` (just a wild guess, I just don't understand what's going on).

**Comment by shunting314** [comment]
> It's odd that those kernels, when fused, have so many spilled registers, when the fusion is pretty much concatenating the code from one after the other... I wonder whether triton keeps some registers alive for longer than it should just because we are marking the loads with`evict_last` (just a wild guess, I just don't understand what's going on).

Hmm, during benchmarking fusion, we have more aggressive using of 'evict_last' since we assume each node's last_usage is empty: i.e., all the buffers will be considered being used later. The reason is 'last_usage' is some global information that's only known after all the fusions are done. But I think this only affect the behavior of cache and the effect should be subtle in most cases.

**Comment by peterbell10** [comment]
It's down to the block sizes chosen for reduction kernels. The unfused pointwise kernel kernel has 2 elements per thread but after fusing with the reduction kernel, it goes up to 8 elements per thread which you would expect to use ~4x as many registers so 111 x 4 is indeed more than 255 and we spill.

**Comment by lezcano** [comment]
That makes sense. Now, I wonder why are we just using 2 elements per thread rather than 4. I would assume we want to use 4-element vectorised loads...

I guess that if it's possible to generate a kernel with plenty of registers that just uses 2 elements per thread, we should use that when taking into consideration fusing it with a kernel that will use 4 elements per thread. Now, this problem would go away if all the kernels of a given size processed the same amount of elements per thread.

**Comment by shunting314** [comment]
> It's down to the block sizes chosen for reduction kernels. The unfused pointwise kernel kernel has 2 elements per thread but after fusing with the reduction kernel, it goes up to 8 elements per thread which you would expect to use ~4x as many registers so 111 x 4 is indeed more than 255 and we spill.

This makes sense. And I realized my previous work that scaling down rblock when there is register pressure does not handle the register spill here due to a loose upper bound. Here is a improvement: https://github.com/pytorch/pytorch/pull/109839 . The spilled kernel now has 2x speedup

**Comment by jon-chuang** [comment]
> The unfused pointwise kernel kernel has 2 elements per thread but after fusing with the reduction kernel, it goes up to 8 elements per thread which you would expect to use ~4x as many registers so 111 x 4 is indeed more than 255 and we spill.

Does Triton expose an estimate of register usage, or is there any tool that can expose the estimated register usage from the generated PTX, so that it can be fed back into the fusion decision?

**Comment by lezcano** [comment]
@jon-chuang triton does expose the number of spilled registers and used regs for a kernel. In https://github.com/pytorch/pytorch/pull/90994 I wrote an upper bound esitmation for the number of registers, but never ended up using it for anything and didn't merge it

**Comment by shunting314** [comment]
@jansel  I updated the PR to skip a fusion choice if the fusion will cause register spills. It's less likely, but if any input node for a fusion has spilled registers, we will skip the fusion as well. Please take another look. Thanks

File: torch/_inductor/codegen/common.py [hunk 2]
```diff
@@ -428,12 +428,10 @@ def __init__(self, name, line):
...
-        if (
-            self.name
-            not in (
-                V.graph.removed_buffers | getattr(V.kernel, "removed_buffers", set())
-            )
-            and self.name not in V.graph.inplaced_to_remove
+        if self.name not in (
+            V.graph.removed_buffers | getattr(V.kernel, "removed_buffers", set())
+        ) and self.name not in (
+            V.graph.inplaced_to_remove | getattr(V.kernel, "inplaced_to_remove", set())
```
Comment (jansel) [inline torch/_inductor/codegen/common.py:434]: It would be cleaner if V.kernel.removed_buffers/inplaced_to_remove were just always defined, so we could read them unconditionally.

This getattr() with defaults pattern can lead to silent correctness issues (rather than loud errors) if someone typos the name.

File: torch/_inductor/codegen/common.py [hunk 3]
```diff
@@ -635,7 +633,16 @@ def aliases(self):
...
-                if other in V.graph.inplaced_to_remove:
+                # For cpu codegen, aliases method may be called in CppScheduling.flush
+                # where we can no longer access the kernel object. It does not matter
+                # that much since the inplaced_to_remove object recorded in the
+                # kernel object should have already being merged to
+                # V.graph.inplaced_to_remove.
+                # We just need handle the case that V.kernel is null gracefully here.
+                if other in (
+                    V.graph.inplaced_to_remove
+                    | getattr(V.kernel, "inplaced_to_remove", set())
```
Comment (jansel) [inline torch/_inductor/codegen/common.py:644]: Same

File: torch/_inductor/config.py [hunk 4]
```diff
@@ -29,6 +29,7 @@
 
 # put correctness assertions in generated code
 size_asserts = os.environ.get("TORCHINDUCTOR_SIZE_ASSERTS", "1") == "1"
+nan_asserts = os.environ.get("TORCHINDUCTOR_NAN_ASSERTS") == "1"
```
Comment (jansel) [inline torch/_inductor/config.py:32]: Let's move this to a different PR.

File: benchmarks/dynamo/common.py [hunk 5]
```diff
@@ -3170,7 +3170,24 @@ def process_entry(rank, runner, original_dir, args):
...
+    """
+    import os
+
+    filename = os.environ.get("PYTHONSTARTUP")
+    if filename and os.path.isfile(filename):
+        with open(filename) as fobj:
+            startup_file = fobj.read()
+        exec(startup_file)
+
+
```
Comment (jansel) [inline benchmarks/dynamo/common.py:3188]: This should be a different PR.  I'm a bit skeptical of this idea... why can't you just set env variables in bash scripts?

**Review by jansel — CHANGES_REQUESTED** [review]

File: benchmarks/dynamo/common.py (hunk 5 above)
Comment (shunting314) [reply benchmarks/dynamo/common.py:3188]: yea, in most cases we can use a shell script wrapper. The slight advantage of the python startup script is

1. we can override some configs that does not have corresponding env variables yet. (although we can add environment variables for them as an alternative)
2. I feel python script more flexible and easier to manage.

But those are not very big advantage TBH, so I'm fine to drop it

File: torch/_inductor/codegen/common.py (hunk 2 above)
Comment (shunting314) [reply torch/_inductor/codegen/common.py:434]: Just recall the reason why I call getattr with default value in the first place:
- I already initialize `removed_buffers` and `inplaced_to_remove` to `set()` in the base kernel class `Kernel`
- However this method may be called outside of a kernel context. i.e., `V.kernel` is a `NullHandler`. The default value for `getattr` is mainly for this.

To resolve the concern regarding getattr with default value, I'll create a NullKernelHandler with `removed_buffers` and `inplaced_to_remove` properly initialized to `set()`. And use NullKernelHandler (rather than NullHandler) if there is no real kernel in the context.

File: torch/_inductor/codegen/common.py [hunk 6]
```diff
@@ -438,14 +438,10 @@ def __init__(self, name, line):
...
-        if (
-            self.name
-            not in (
-                V.graph.removed_buffers | getattr(V.kernel, "removed_buffers", set())
-            )
-            and self.name not in V.graph.inplaced_to_remove
+        if self.name not in (
+            V.graph.removed_buffers | V.kernel.removed_buffers
+        ) and self.name not in (
+            V.graph.inplaced_to_remove | V.kernel.inplaced_to_remove
```
Comment (jansel) [inline torch/_inductor/codegen/common.py:444]: `self.name not in V.graph.removed_buffers and name not in V.kernel.removed_buffers` would be faster since it is O(1) rather than O(n).  There are two similar cases below.

**Review by jansel — APPROVED** [review]

File: torch/_inductor/codegen/triton.py [hunk 7]
```diff
@@ -2825,6 +2830,81 @@ def select_tiling(cls, node_schedule, numel, reduction_numel=sympy.Integer(1)):
...
+        wrapped_jit_function = mod.triton_
+
+        # call once to trigger the compilation
+        call(wrapped_jit_function.clone_args(*args))
+
+        launchers = wrapped_jit_function.launchers
+        assert len(launchers) == 1
+        if launchers[0].n_spills > 0:
+            # skip benchmarking the kernel if there are register spills
+            ms = float("inf")
```
Comment (lezcano) [inline torch/_inductor/codegen/triton.py:2894]: Sometimes using spills is not actually bad, so perhaps we want to remove this optimisation?

Comment (lezcano) [reply torch/_inductor/codegen/triton.py:2894]: Instead, it may be better to simply log that there were register spills and how many

Comment (shunting314) [reply torch/_inductor/codegen/triton.py:2894]: hmm, do you have examples that the fused spilled kernel can run faster (when the kernels before fusion does not spill)? I assume even if we have such examples, it would be really rare? So the optimization would still be a good heuristic to prune unnecessary benchmarking work.

Comment (lezcano) [reply torch/_inductor/codegen/triton.py:2894]: I don't have any examples, but I think it could happen if the number of spilled registers is not too large. Registers first spill into L1, so if there aren't that many and they fit in L1, spilling registers is not particularly slow. This can get even better if we use cache modifiers like `.cg` when we know we are not going to reuse the data, as to not pollute the L1.

Perhaps we could change this to still benchmark if there are not too many spilled registers?

Comment (shunting314) [reply torch/_inductor/codegen/triton.py:2894]: I agree that cache can help a bit to make accessing the spilled register not too slow. Although I don't know how helpful that can be since GPU L1 cache is relatively small.

But more importantly when register start spills, each thread usually already uses 255 registers (the max #of registers a thread can use). That already limit the occupancy to be at most 12.5% .  It's not always but such low occupancy can usually cause bad perf.

**Comment by ZainRizvi** [comment]
@pytorchbot revert -m "This test is breaking trunk. In the future please make sure to add the ciflow/trunk label before force merging any PR to ensure your code doesn't break those tests"

https://hud.pytorch.org/pytorch/pytorch/commit/ec0cdcdf6a816eadb4d868284eea86732f50da2e

**Comment by ZainRizvi** [comment]
@pytorchbot revert -m "This test is breaking trunk. In the future please make sure to add the ciflow/trunk label before force merging any PR to ensure your code doesn't break those tests" -c ignoredsignal

https://hud.pytorch.org/pytorch/pytorch/commit/ec0cdcdf6a816eadb4d868284eea86732f50da2e

**Comment by shunting314** [comment]
> @pytorchbot revert -m "This test is breaking trunk. In the future please make sure to add the ciflow/trunk label before force merging any PR to ensure your code doesn't break those tests" -c ignoredsignal
> 
> https://hud.pytorch.org/pytorch/pytorch/commit/ec0cdcdf6a816eadb4d868284eea86732f50da2e

@ZainRizvi the failed check seems to be only performed at land time. Is the ciflow/trunk label you mentioned force running them earlier (e.g. before merging is kicked off)?
